http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/IdleConnectionCountingChannelPool.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty.internal;
+
+import static software.amazon.awssdk.http.nio.netty.internal.utils.NettyUtils.doInEventLoop;
+
+import io.netty.channel.Channel;
+import io.netty.channel.pool.ChannelPool;
+import io.netty.util.AttributeKey;
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.Promise;
+import java.util.concurrent.CompletableFuture;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.http.HttpMetric;
+import software.amazon.awssdk.metrics.MetricCollector;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * A channel pool implementation that tracks the number of "idle" channels in an underlying channel pool.
+ *
+ * <p>Specifically, this pool counts the number of channels acquired and then released from/to the underlying channel pool. It
+ * will monitor for the underlying channels to be closed, and will remove them from the "idle" count.
+ */
+@SdkInternalApi
+public class IdleConnectionCountingChannelPool implements SdkChannelPool {
+    private static final Logger log = Logger.loggerFor(IdleConnectionCountingChannelPool.class);
+
+    /**
+     * The idle channel state for a specific channel. This should only be accessed from the {@link #executor}.
+     */
+    private static final AttributeKey<ChannelIdleState> CHANNEL_STATE =
+        AttributeKey.newInstance("IdleConnectionCountingChannelPool.CHANNEL_STATE");
```

Comment:
should we namespace the key?

Reply:
What do you mean?

Reply:
I meant prefixing it with sdk namespace so it doesn't conflict with other keys. It seems we are using `NettyUtils.getOrCreateAttributeKey` now introduced in #1887